warehouse-management-service/pkg/database/postgres/shelf.go

```diff
@@ -0,0 +1,198 @@
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	wms "warehouse-management-service"
+)
+
+type shelfQueries interface {
+	createShelfTx(ctx context.Context, tx *sql.Tx, shelf wms.Shelf) error
+	getShelfByIdTx(ctx context.Context, tx *sql.Tx, id string) (wms.Shelf, error)
+	updateShelfTx(ctx context.Context, tx *sql.Tx, shelf wms.Shelf) error
+	deleteShelfTx(ctx context.Context, tx *sql.Tx, id string) error
+	shelfBlockExistsTx(ctx context.Context, tx *sql.Tx, id string) (bool, error)
+}
+
+type shelfQueriesImpl struct{}
+
+type ShelfService struct {
+	queries shelfQueries
+	db      *sql.DB
+}
+
+func NewShelfService(db *sql.DB) *ShelfService {
+	return &ShelfService{
+		queries: new(shelfQueriesImpl),
+		db:      db,
+	}
+}
+
+var InvalidShelfBlock = errors.New("invalid shelfBlockId")
+
+func (s *ShelfService) GetShelfById(ctx context.Context, id string) (wms.Shelf, error) {
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return wms.Shelf{}, err
+	}
+
+	shelf, err := s.queries.getShelfByIdTx(ctx, tx, id)
+	switch err {
+	case nil:
+		return shelf, tx.Commit()
+	case sql.ErrNoRows:
+		return wms.Shelf{}, wms.ShelfDoesNotExist
+	default:
+		return wms.Shelf{}, err
+	}
+}
+
+func (s *ShelfService) CreateShelf(ctx context.Context, shelf wms.Shelf) error {
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	err = s.queries.createShelfTx(ctx, tx, shelf)
+	switch err {
+	case InvalidShelfBlock:
+		return wms.InvalidShelfBlock
+	case nil:
+		return tx.Commit()
+	default:
+		return err
+	}
+}
+
+func (s *ShelfService) UpdateShelf(ctx context.Context, shelf wms.Shelf) error {
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	err = s.queries.updateShelfTx(ctx, tx, shelf)
+	switch err {
+	case nil:
+		return tx.Commit()
+	case RowDoesNotExist:
+		return wms.ShelfDoesNotExist
+	case InvalidShelfBlock:
+		return wms.InvalidShelfBlock
+	default:
+		return err
+	}
+}
+
+func (s *ShelfService) DeleteShelfById(ctx context.Context, id string) error {
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	err = s.queries.deleteShelfTx(ctx, tx, id)
+	switch err {
+	case nil:
+		return tx.Commit()
+	case RowDoesNotExist:
+		return wms.ShelfDoesNotExist
+	default:
+		return err
+	}
+}
+
+func (s *shelfQueriesImpl) createShelfTx(ctx context.Context, tx *sql.Tx, shelf wms.Shelf) error {
+	if shelfBlockExists, err := s.shelfBlockExistsTx(ctx, tx, shelf.ShelfBlockId); err != nil {
```

Comment:
Can this logic can be moved from code to database by virtue of foreign keys?
In some high-scale applications, FKs are considered detrimental to performance. 
However, in this case, we can leverage them.

Reply:
The reasoning behind having this logic was to be able to distinguish between querying a non-existent record vs a generic internal error. However, this explicit check might be excessive and I am considering changing this to check for foreign key errors returned by the library.